### PR TITLE
BMD-13509 Editing existing permissions

### PR DESCRIPTION
**Pull request by bmxarch**
### Jira ticket:
https://berlingskemedia.atlassian.net/browse/BMD-13509

### CODE REVIEW

File: client/src/Main.tsx [hunk 1]
```diff
@@ -37,6 +38,10 @@ export function Main(): JSX.Element {
           path="/newsletters/create-new-permission"
           component={PermissionCreator}
         />
+        <Route
+          path="/newsletters/edit-permission"
```
Comment (bmxdaba) [inline client/src/Main.tsx:42]: you cannot open a specifc newsletter using URL only. Change to `/newsletters/permission/:id`

File: client/src/components/FormField/Select.tsx [hunk 2]
```diff
@@ -37,7 +42,8 @@ const Select: React.FC<SelectProps> & { Option: typeof Option } = ({
           setCurrentValue(event.target.value);
           selectProps.onChange?.(event);
         }}
-        value={currentValue || ''}>
+        value={currentValue || ''}
```
Comment (bmxdaba) [inline client/src/components/FormField/Select.tsx:45]: I would move it here
```typescript
  const [currentValue, setCurrentValue] = React.useState(selectProps.value || '');
```

File: client/src/newsletter/Permission.tsx [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
 import styles from './permission.module.scss';
 import { BrandsFullNames } from './PermissionsFilter';
+import { PermissionInfo } from 'newsletter/PermissionEditor';
 
 type OwnProps = {
   permission: PermissionType;
 };
 
 export const Permission = ({ permission }: OwnProps): JSX.Element => {
+  const permissionInfo = {
```
Comment (bmxdaba) [inline client/src/newsletter/Permission.tsx:18]: note: extract a function to map the value; just for readability

**Review by bmxdaba — CHANGES_REQUESTED** [review]

File: client/src/newsletter/PermissionEditor.tsx [hunk 4]
```diff
@@ -106,17 +114,48 @@ export const PermissionEditor = (): JSX.Element => {
...
+          consentText,
+        } as FormValues;
+
+        setPermissionData(permissionData);
+      } catch (e) {
+        setAlert('The permission with the given id does not exist!');
+        console.error(e);
+      } finally {
+        setIsPending(false);
+      }
```
Comment (bmxdaba) [inline client/src/newsletter/PermissionEditor.tsx:138]:
```typescript
// outside the component
const mapPermissionToFormValues = (permission: Permission): FormValues => ({
   // map values
});

//then.. 
      try {
        setIsPending(true);
        const permission = await newsletterSdk.permissions.admin.get(id);
        setPermissionData(mapPermissionToFormValues(permission));
      } catch (error) {
        setAlert('The permission with the given id does not exist!');
        console.error(error);
      } finally {
        setIsPending(false);
      }
```

File: client/src/newsletter/PermissionEditor.tsx [hunk 5]
```diff
@@ -51,20 +52,32 @@ const checkChanges = (
...
+        Back
+      </button>
+    </div>
+  );
+};
+
 export const PermissionEditor = (): JSX.Element => {
-  const { state: locationState } = useLocation<PermissionInfo | undefined>();
+  const { id } = useParams<ReactRouterParams>();
+  const [permissionData, setPermissionData] = React.useState<FormValues | null>(
```
Comment (bmxdaba) [inline client/src/newsletter/PermissionEditor.tsx:73]: I would rename it to `const [formValues, setFormValues]`

**Review by bmxdaba — CHANGES_REQUESTED** [review]

File: client/src/newsletter/PermissionEditor.tsx (hunk 4 above)
Comment (bmxdaba) [reply client/src/newsletter/PermissionEditor.tsx:138]: and here `setAlert('The permission with the given id does not exist!');` should only be displayed for `404` statusCode

so it should be like this:

```typescript
catch (error) {
   const errorMessage = error.statusCode === 404
             ? 'The permission with the given id does not exist!'
             : error.message;
   setAlert(errorMessage);
   console.error(error);
}
```

File: client/src/newsletter/PermissionEditor.tsx (hunk 5 above)
Comment (bmxarch) [reply client/src/newsletter/PermissionEditor.tsx:73]: I think `[initFormValues, setInitFormValues]` will be best option.

This fragment is more readable in my opinion
```
const [initFormValues, setInitFormValues]
...
...
 const updatePermission = async (formValues: FormValues): Promise<void> => {
    if (!initFormValues) return;
    try {
      setIsPending(true);
      if (!checkChanges(formValues, initFormValues)) {
        setAlert('No changes was provided!');
        throw new Error('validation');
      }
```

File: client/src/newsletter/PermissionEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+          alert={alert}
+          isProcessing={isPending}
+          requiredFields={requiredFields}
+          disabledElements={disabledElements}
+          handleCancel={() => handleSwitchRoute('cancel')}
+          onSubmit={updatePermission}
+          handleResetAlert={handleResetAlert}
+        />
+      )}
+      {!initFormValues && alert && <PermissionNotFound alertMsg={alert} />}
```
Comment (bmxdaba) [inline client/src/newsletter/PermissionEditor.tsx:183]: note: 
```suggestion
      {initFormValues ? (
        <PermissionForm
          submitButtonDisplayValue="Save"
          initFormValues={initFormValues}
          alert={alert}
          isProcessing={isPending}
          requiredFields={requiredFields}
          disabledElements={disabledElements}
          handleCancel={() => handleSwitchRoute('cancel')}
          onSubmit={updatePermission}
          handleResetAlert={handleResetAlert}
        />
      ) : alert ? (
        <PermissionNotFound alertMsg={alert} />
      ) : null}
```

**Review by bmxdaba — CHANGES_REQUESTED** [review]

File: client/src/newsletter/PermissionEditor.tsx (hunk 5 above)
Comment (bmxdaba) [reply client/src/newsletter/PermissionEditor.tsx]: but they aren't init form values, just a form values :thinking:

**Review by bmxdaba — APPROVED** [review]

File: client/src/newsletter/PermissionEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+): UpdatePermissionDto => {
+  return {
+    ...values,
+    brandId: parseInt(values.brandId),
+    sortId: 0,
+    permission: { connect: { permissionId } },
+  };
+};
+
+const hasFormValuesChanged = (
```
Comment (bmxdaba) [inline client/src/newsletter/PermissionEditor.tsx:49]: `have`; my bad :sweat_smile:

**Review by bmxpiku — APPROVED** [review]

**Review by bmxpiku — APPROVED** [review]

**Pull request merged**